zetaclient/cctx_scanner.go

```diff
@@ -0,0 +1,229 @@
+		sc.missedPendingCctx[chainID][nonce] = send
+	}
+	sc.nextNonceToScan[chainID] = nonceTo
+	if len(missedList) > 0 {
+		sc.saveFirstNonceToScan(chainID)
+		sc.logger.Info().Msgf("scanner: found missed pending cctx for chain %d with nonces %v", chainID, nonces)
+	}
+}
+
+func (sc *CctxScanner) removeMissedPendingCctx(chainID int64, nonce uint64) {
+	delete(sc.missedPendingCctx[chainID], nonce)
+	sc.saveFirstNonceToScan(chainID)
+}
+
+func (sc *CctxScanner) saveFirstNonceToScan(chainID int64) {
+	firstNonceToScan := uint64(math.MaxUint64)
+	if len(sc.missedPendingCctx[chainID]) == 0 {
+		// either no missed pending cctx found so far OR last missed pending cctx removed
+		firstNonceToScan = sc.nextNonceToScan[chainID]
+	} else { // save the lowest nonce for future restart if there ARE missed pending cctx
+		for nonceMissed := range sc.missedPendingCctx[chainID] {
+			if nonceMissed < firstNonceToScan {
+				firstNonceToScan = nonceMissed
+			}
+		}
+	}
+	if firstNonceToScan < uint64(math.MaxUint64) {
+		if err := sc.db.Save(clienttypes.ToFirstNonceToScanSQLType(chainID, firstNonceToScan)).Error; err != nil {
+			sc.logger.Error().Err(err).Msgf("scanner: error writing firstNonceToScan for chain %d nonce %d", chainID, firstNonceToScan)
+		}
+	}
+}
+
+// LoadDB open sql database and load data into scanner
+func (sc *CctxScanner) LoadDB(dbpath string, memDB bool) error {
+	if _, err := os.Stat(dbpath); os.IsNotExist(err) {
+		err := os.MkdirAll(dbpath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+	path := dbpath
+	if !memDB { // memDB is used for uint test only
+		path = fmt.Sprintf("%s/scanner", dbpath)
+	}
+	db, err := gorm.Open(sqlite.Open(path), &gorm.Config{})
+	if err != nil {
+		return err
+	}
+	sc.db = db
+
+	err = db.AutoMigrate(&clienttypes.CurrentTssSQLType{},
+		&clienttypes.FirstNonceToScanSQLType{})
+	if err != nil {
+		return err
+	}
+
+	// Load current tss pubkey
+	sc.loadCurrentTssPubkey()
+
+	// Load first nonce for each chain to start scanning from
+	err = sc.buildFirstNonceToScanMap()
+
+	return err
+}
+
+func (sc *CctxScanner) Reset(tssPubkey string) error {
+	sc.tssPubkey = tssPubkey
+	sc.nextNonceToScan = make(map[int64]uint64)
+	sc.missedPendingCctx = make(map[int64]map[uint64]*crosschaintypes.CrossChainTx)
+
+	// save current tss pubkey
+	if err := sc.db.Save(clienttypes.ToCurrentTssSQLType(tssPubkey)).Error; err != nil {
+		sc.logger.Error().Err(err).Msgf("scanner: error writing current tss pubkey %s", tssPubkey)
+		return err
+	}
+
+	// clean db, GORM uses pluralizes struct name to snake_cases as table name
+	if err := sc.db.Exec("DELETE FROM first_nonce_to_scan_sql_types").Error; err != nil {
+		sc.logger.Error().Err(err).Msg("scanner: error cleaning FirstNonceToScan db")
+		return err
+	}
+	sc.logger.Info().Msgf("scanner: reset db successfully for tss pubkey %s", tssPubkey)
+
+	return nil
+}
+
+func (sc *CctxScanner) loadCurrentTssPubkey() {
```

Comment:
I think it would be more maintainable to return the errors in the low-level functions like this one and printing the error in the high level function.

Not a blocker

Reply:
Agreed. That's the ideal error handling.